### PR TITLE
Convert Binary tree to Doubly linked list

### PR DESCRIPTION
**Pull request by Ashee-Kanungo**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

I have added the code in c++ to convert binary tree to doubly linked list.

### Checklist

- [Yes] I've read the contribution guidelines.
- [Yes] I've checked the issue list before deciding what to submit.
- [Yes] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#2754

### CODE REVIEW

**Comment by Ashee-Kanungo** [comment]
@ankitaggarwal23 @gat786 @anushkrishnav @siddharth25pandey @HarshCasper please review this PR.

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
+/*
+This code gets the values of binary tree from the user and then convert the binary tree to doubly linked list.
+The formed Doubly linked list will have elements in the same order as that of inorder traversal of binary tree. 
+*/
+
+#include <iostream>
+#include <vector>
+#include <queue>
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Replace the header files with `#include<bits/stdc++.h>`

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+};
+
+TreeNode* newNode(int n){
+	TreeNode* t = new TreeNode;
+	t->value = n;
+	t->left = NULL;
+	t->right = NULL;
+}
+
+TreeNode* insert(TreeNode* t, queue<TreeNode* > &q, int v){		//This function inserts node into the binary tree
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Shift the comment above the function

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	else{
+		q.front()->right = temp;
+		q.pop();
+	}
+	q.push(temp);
+	return t;
+}
+
+TreeNode* make_tree(vector<int> v){		//This function converts given vector into Binary Tree
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Shift the comment above the function

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	queue<TreeNode* > q;
+	TreeNode* t = NULL;
+	for(int i=0;i<v.size();i++){
+		t = insert(t,q,v[i]);
+	}
+	return t;
+}
+
+void print_elements(TreeNode* &T){		//This function prints all the elements of doubly linked list
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Shift the comment above the function

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	TreeNode* temp = T;
+	cout<<"The elements of doubly linked list are: ";
+	while(temp!=NULL){
+		cout<<temp->value<<" ";
+		temp = temp->right;
+	}
+}
+
+void to_dll(TreeNode* T, TreeNode* &temp){		//This function converts Binary tree to doubly linked list
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Shift the comment above the function

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+The output will be - 3 6 4 5 9 8 2
+
+Suppose if input binary tree is:
+    9
+   / \
+  8   2
+ /   / 
+5    4 
+The output will be - 5 8 9 4 2
+*/
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: - Remove the extra testcase one is enough. 
- Add Time and Space Complexity at the end of the Code   
- Add new line at the EOF

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by Ashee-Kanungo** [comment]
@ankitaggarwal23 I have made all the changes in the PR. Please review it.

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+void to_doubly_linked_list(TreeNode* T){		
+	TreeNode* h = NULL;
+	to_dll(T,h);
+	print_elements(h);
+}
+
+int main(){
+	int n;		//number of nodes of binary tree
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Write a good naming convention for n such as nodes

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	cin>>n;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Before this line, please print the the message as what are you taking input from the user for better UI

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	vector<int> v;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: write a good naming convention for v

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	while(n){		//asking user for elements of Binary tree
+		int x;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: write a good naming convention for x

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+		cin>>x;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Before this line, please print the the message as what are you taking input from the user.

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+		v.push_back(x);
+		n--;
+	}
+	TreeNode* T = make_tree(v);
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: Before this line, please print the the message as what are you showing to the user

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+	to_doubly_linked_list(T);
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp:102]: Before this line, please print the the message as what are you showing to the user

File: C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+Here if we give following binary tree as input:
+    5
+   / \
+  6   8
+ / \ / \
+3  4 9  2 
+The output will be - 3 6 4 5 9 8 2
+*/
+
```
Comment (siddharth25pandey) [inline C-Plus-Plus/cp/Binary_Tree_to_Doubly_Linked_List.cpp]: In Sample Input & Output, Display exactly same as what your program is executing

**Review by siddharth25pandey — CHANGES_REQUESTED** [review]
@Ashee-Kanungo Do the suggested changes

**Comment by siddharth25pandey** [comment]
@Ashee-Kanungo Also resolve the merge conflict error

**Comment by Ashee-Kanungo** [comment]
@siddharth25pandey how can I resolve the merge conflict error? Should I make a separate PR with all these changes?

**Comment by siddharth25pandey** [comment]
> @siddharth25pandey how can I resolve the merge conflict error? Should I make a separate PR with all these changes?

@Ashee-Kanungo I have resolved the merge conflict error. Only you need to modify the changes as suggested earlier in the same PR

**Comment by Ashee-Kanungo** [comment]
@siddharth25pandey  I have made all the changes in the PR. Please review it.

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Write output what your code is producing.

**Comment by Ashee-Kanungo** [comment]
@ankitaggarwal23 I have made the changes. Please review it.

**Review by ankitaggarwal23 — COMMENTED** [review]
I have checked your code and it shows error kindly check your code and also write the output in the same way the code is producing.

**Comment by Ashee-Kanungo** [comment]
@ankitaggarwal23 I have made the changes. Please review it.

**Review by ankitaggarwal23 — APPROVED** [review]

**Pull request merged**